autoPyTorch/utils/hyperparameter_search_space_update.py

```diff
@@ -29,30 +55,67 @@ def __str__(self) -> str:
     def __init__(self, updates: Optional[List[HyperparameterSearchSpaceUpdate]] = None) -> None:
         self.updates = updates if updates is not None else []
 
     def apply(self, pipeline: List[Tuple[str, Union[autoPyTorchComponent, autoPyTorchChoice]]]) -> None:
+        """
+        Iteratively applies updates to the pipeline
+        Args:
+            pipeline: (List[Tuple[str, Union[autoPyTorchComponent, autoPyTorchChoice]]]):
+                The named steps of the current autopytorch pipeline
+        Returns:
+            None
+        """
         for update in self.updates:
             update.apply(pipeline)
 
     def append(self, node_name: str, hyperparameter: str, value_range: Union[List, Tuple],
                default_value: Union[int, float, str], log: bool = False) -> None:
+        """
+        Add a new update
+        Args:
+            node_name (str):
+            The name of the node in the pipeline
+        hyperparameter (str):
+            The name of the hyperparameter
+        value_range (Union[List, Tuple]):
+            In case of categorical hyperparameter, defines the new categorical choices.
+            In case of numerical hyperparameter, defines the new range
+            in the form of (LOWER, UPPER)
+        default_value (Union[int, float, str]):
+            New default value for the hyperparameter
+        log (bool) (default=False):
+            In case of numerical hyperparameters, whether to sample on a log scale
+
+        Returns:
+            None
+        """
         self.updates.append(HyperparameterSearchSpaceUpdate(node_name=node_name,
                                                             hyperparameter=hyperparameter,
                                                             value_range=value_range,
                                                             default_value=default_value,
                                                             log=log))
 
     def save_as_file(self, path: str) -> None:
+        """
+        Save the updates as a file to reuse later
+        Args:
+            path (str): path of the file
+
+        Returns:
+            None
+        """
         with open(path, "w") as f:
-            with open(path, "w") as f:
-                for update in self.updates:
-                    print(update.node_name, update.hyperparameter,  # noqa: T001
-                          str(update.value_range), "'{}'".format(update.default_value)
-                          if isinstance(update.default_value, str) else update.default_value,
-                          (" log" if update.log else ""), file=f)
+            for update in self.updates:
+                print(update.node_name, update.hyperparameter,  # noqa: T001
+                      str(update.value_range), "'{}'".format(update.default_value)
+                      if isinstance(update.default_value, str) else update.default_value,
+                      (" log" if update.log else ""), file=f)
```

Comment:
From next time, let's separate the PR for refactoring:)

Reply:
this was a bug I found as we are opening the file twice. As it is not a major update I just added to this PR

Reply:
Ok, got it